### PR TITLE
TextEditor: Clear leftover whitespace when inserting newlines

### PR DESCRIPTION
**Pull request by euclidianAce**
Previously when entering a newline, previous indentation would be left,
leaving a line consisting only of whitespace. This fixes that

When watching Andreas build something inside of Serenity itself, it always irked me that there was a not insignificant amount of time correcting whitespace leftover by the autoindent. Hopefully this fixes that :^)

I haven't delved deep into the code for the TextEditor widget, so if there's a better way of doing this let me know!

### CODE REVIEW

File: Userland/Libraries/LibGUI/TextEditor.cpp [hunk 1]
```diff
@@ -1244,7 +1244,23 @@ void TextEditor::insert_at_cursor_or_replace_selection(const StringView& text)
     VERIFY(is_editable());
     if (has_selection())
         delete_selection();
+
+    // check if adding a newline leaves the previous line as just whitespace
+    auto const clr_len = m_cursor.column();
+    auto const should_clear_last_ln = text == "\n"
```
Comment (awesomekling) [inline Userland/Libraries/LibGUI/TextEditor.cpp:1250]: Please write out full words in variable names instead of using random abbreviations, unless the abbreviation makes it significantly clearer.

File: Userland/Libraries/LibGUI/TextEditor.cpp [hunk 2]
```diff
@@ -1244,7 +1244,23 @@ void TextEditor::insert_at_cursor_or_replace_selection(const StringView& text)
...
+        && clr_len > 0
+        && current_line().leading_spaces() == clr_len;
+
     execute<InsertTextCommand>(text, m_cursor);
+
+    if (should_clear_last_ln) { // if it does leave just whitespace, clear it
+        auto const orig_cursor_pos = cursor();
```
Comment (awesomekling) [inline Userland/Libraries/LibGUI/TextEditor.cpp:1257]: Same here.

**Review by awesomekling — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibGUI/TextEditor.cpp [hunk 3]
```diff
@@ -1244,7 +1244,23 @@ void TextEditor::insert_at_cursor_or_replace_selection(const StringView& text)
     VERIFY(is_editable());
     if (has_selection())
         delete_selection();
+
+    // check if adding a newline leaves the previous line as just whitespace
```
Comment (linusg) [inline Userland/Libraries/LibGUI/TextEditor.cpp:1248]: Comments should look like sentences, with proper punctuation (https://github.com/SerenityOS/serenity/blob/master/Documentation/CodingStyle.md#comments)

```suggestion
    // Check if adding a newline leaves the previous line as just whitespace.
```

File: Userland/Libraries/LibGUI/TextEditor.cpp [hunk 4]
```diff
@@ -1244,7 +1244,23 @@ void TextEditor::insert_at_cursor_or_replace_selection(const StringView& text)
...
+    auto const clear_length = m_cursor.column();
+    auto const should_clear_last_line = text == "\n"
+        && clear_length > 0
+        && current_line().leading_spaces() == clear_length;
+
     execute<InsertTextCommand>(text, m_cursor);
+
+    if (should_clear_last_line) { // if it does leave just whitespace, clear it
```
Comment (linusg) [inline Userland/Libraries/LibGUI/TextEditor.cpp:1256]: same here

**Review by linusg — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibGUI/TextEditor.cpp [hunk 5]
```diff
@@ -803,7 +803,7 @@ void TextEditor::keydown_event(KeyEvent& event)
         return;
     }
 
-    if (event.key() == KeyCode::Key_Backspace) {
+    if (event.key() == KeyCode::Key_Backspace || event.code_point() == '\b') {
```
Comment (linusg) [inline Userland/Libraries/LibGUI/TextEditor.cpp:806]: Why is this needed?

Comment (euclidianAce) [reply Userland/Libraries/LibGUI/TextEditor.cpp:806]: Ah, that was my local fix for #6748. Didn't mean to commit it with this patch!
